fix: add flag dedicated server override v2.x

noellie-velez opened the pull request:
## Purpose of this PR
[//]: # (
Replace this block with what this PR does and why. Describe what you'd like reviewers to know, how you applied the engineering principles, and any interesting tradeoffs made. 
)

fix: #3490

### Jira ticket
[MTTB-1557](https://jira.unity3d.com/browse/MTTB-1557)

### Changelog
[//]: # (updated with all public facing changes  - API changes, UI/UX changes, behaviour changes, bug fixes. Remove if not relevant.)

- Added: A flag to override command-line arguments (port and ip) in `SetConnectionData`.
- Added: A command-line singleton to parse environment command-line arguments.
- Fixed: Issue where using the dedicated server package would override all attempts to change the port by code.

<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater](https://confluence.unity3d.com/display/DEV/Obsolete+API+updaters) was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

## Documentation
[//]: # (
This section is REQUIRED and should mention what documentation changes were following the changes in this PR. 
We should always evaluate if the changes in this PR require any documentation changes.
)

- Includes new documentation for CommandLineOptions.

## Testing & QA (How your changes can be verified during release Playtest)
[//]: #  (
This section is REQUIRED and should describe how the changes were tested and how should they be tested when Playtesting for the release.
It can range from "edge case covered by unit tests" to "manual testing required and new sample was added".
Expectation is that PR creator does some manual testing and provides a summary of it here.)

<!-- Add any performance testing results here if relevant. -->

### Functional Testing
[//]: # (If checked, List manual tests that have been performed.)
_Manual testing :_
- [x] `Manual testing done`
  - Install the dedicated server package, add script that uses the override parameter and override the port or ip address.

_Automated tests:_
- [ ] `Covered by existing automated tests`
- [ ] `Covered by new automated tests`

_Does the change require QA team to:_

- [ ] `Review automated tests`?
- [ ] `Execute manual tests`?
- [ ] `Provide feedback about the PR`?

If any boxes above are checked the QA team will be automatically added as a PR reviewer.

## Backports

No back port required.

noellie-velez committed:
Moving comand line options to unity transport

noellie-velez committed:
Moving comand line options to unity transport

noellie-velez committed:
Updating project version (can be reverted)

noellie-velez committed:
Disable build scenes + add CommandLineOverrideTest

noellie-velez committed:
Reset NetworkManager and UnityTransport

noellie-velez committed:
Investigate DedicatedServer.Arguments.Port null

noellie-velez committed:
Merge branch 'develop-2.0.0' into fix/v2.x/add-flag-dedicated-server-override

noellie-velez committed:
Merge branch 'develop-2.0.0' into fix/v2.x/add-flag-dedicated-server-override

noellie-velez committed:
Remove non editor upgrade changes

noellie-velez committed:
Package update

noellie-velez committed:
Clean up

noellie-velez commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -785,14 +785,79 @@ public void SetClientRelayData(string ipAddress, ushort port, byte[] allocationI
         /// <param name="listenAddress">The local listen address</param>
-        public void SetConnectionData(string ipv4Address, ushort port, string listenAddress = null)
+        /// <param name="overrideCommandLineArgs">Should override port value</param>
+        public void SetConnectionData(string ipv4Address, ushort port, string listenAddress = null, bool overrideCommandLineArgs = false)
```
rename overrideCommandLineArgs to forceOverrideCommandLineArgs

noellie-velez committed:
Reset/delete files to exlude

noellie-velez committed:
Editor version minimalproject + Example projects

6000.0.61f1

noellie-velez committed:
Editor upgrade testproject 6000.61f1 (latest LTS)

noellie-velez committed:
Merge branch 'develop-2.0.0' into chore/v2.x/testproject-editor-upgrade-6000.0.60f1

noellie-velez committed:
Merge branch 'develop-2.0.0' into fix/v2.x/add-flag-dedicated-server-override

# Conflicts:
#	com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs

noellie-velez committed:
reset manifest

noellie-velez committed:
Merge branch 'chore/v2.x/testproject-editor-upgrade-6000.0.60f1' into fix/v2.x/add-flag-dedicated-server-override

noellie-velez committed:
Rename forceOverrideCommandLineArgs

noellie-velez committed:
Remove Override and ParseArgs, add flag to overide

noellie-velez committed:
Remove logs

noellie-velez committed:
Merge branch 'develop-2.0.0' into fix/v2.x/add-flag-dedicated-server-override

noellie-velez committed:
Address review comments TODO cleanup

noellie-velez committed:
Revert files

noellie-velez committed:
Fix unknown arg when not dedicated server profile

EmandM reviewed (approved):
Looks great! 🚀

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1582,9 +1629,13 @@ public override void Initialize(NetworkManager networkManager = null)
+                {
+                    Debug.Log($"Already has command line option set. Using connection data set to {ConnectionData.Address}:{port}");
+                }
+                ConnectionData.Port = (ushort)port;
```
```suggestion
                ConnectionData.Port = port;
```

This will fix your standards check error. ParseCommandLineOptions is already defining the type written to in the 1st out parameter.

If you opt to make a more generic one, then you might adjust ParseCommandLineOptions to be something like ParseCommandLineOptions<T>.

noellie-velez replied:
Good catch, I missed that one.

noellie-velez committed:
Merge branch 'develop-2.0.0' into fix/v2.x/add-flag-dedicated-server-override

noellie-velez committed:
Revert "Revert files" + test project

This reverts commit 05b794c6e44f6034eb9549f93f0aea5f172d98d2.

noellie-velez committed:
Creating a command line option class + singleton

noellie-velez committed:
Remove DGS and PortOverrideCLI testing files

This reverts commit 4f1217686602da4387b4d958d3039215cc61b953.

noellie-velez committed:
Remove redundant ushort cast

Co-authored-by: Noel Stephens <noel.stephens@unity3d.com>

noellie-velez committed:
Fix maxnumberplayers documentation Note

noellie-velez committed:
Adding documentation to public API

noellie-velez committed:
WIP documentation

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
```diff
@@ -0,0 +1,7 @@
+# WORK IN PROGRESS
+# Command line arguments
```
```suggestion
# Command-line arguments
```

jabbacakes commented:
Looks like the manual doc here isn't finished - do you need a hand with it? If not, just gimme a ping to review it when you've written a bit more.

noellie-velez committed:
Reset testproject files

noellie-velez committed:
Adding returns to get args summary

noellie-velez committed:
Notes for docs, still WIP

noellie-velez committed:
Bumped the package version (job failing issue)

noellie-velez committed:
Update docs

noellie-velez committed:
Added -ip command line option

noellie-velez committed:
Merge branch 'develop-2.0.0' into fix/v2.x/add-flag-dedicated-server-override

# Conflicts:
#	com.unity.netcode.gameobjects/Documentation~/TableOfContents.md
#	com.unity.netcode.gameobjects/Documentation~/configuration.md

noellie-velez commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -805,14 +805,60 @@ public void SetClientRelayData(string ipAddress, ushort port, byte[] allocationI
+                port = commandLinePort;
+            }
+
+            if (ParseCommandLineOptionsAddress(out var commandLineIp))
```
```suggestion
            if (!forceOverrideCommandLineArgs && ParseCommandLineOptionsAddress(out var commandLineIp))
```

noellie-velez committed:
Update to check force also for address

@NoelStephensUnity suggestion

NoelStephensUnity commented on com.unity.netcode.gameobjects/package.json:
```diff
@@ -2,7 +2,7 @@
     "displayName": "Netcode for GameObjects",
     "description": "Netcode for GameObjects is a high-level netcode SDK that provides networking capabilities to GameObject/MonoBehaviour workflows within Unity and sits on top of underlying transport layer.",
-    "version": "2.8.0",
+    "version": "2.9.0",
```
_(Reminder)_
Minor version update bc of CommandLineOptions

noellie-velez committed:
Update changelog and docs

noellie-velez committed:
Merge branch 'fix/v2.x/add-flag-dedicated-server-override' of github.com:Unity-Technologies/com.unity.netcode.gameobjects into fix/v2.x/add-flag-dedicated-server-override

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity reviewed (approved):
Looks good Noellie!
👍

NoelStephensUnity committed:
fix

Convert the string to ushort.

NoelStephensUnity committed:
sty;e

Simplify the ushort conversion using ushort.

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/TableOfContents.md:
```diff
@@ -19,6 +19,7 @@
     * [Max players](basics/maxnumberplayers.md)
   * [Transports](advanced-topics/transports.md)
   * [Unity Relay](relay/relay.md)
+  * [Command line arguments](command-line-arguments.md)
```
```suggestion
  * [Command-line arguments](command-line-arguments.md)
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
```diff
@@ -0,0 +1,63 @@
+# Command line arguments
+
+You can use [command line arguments](https://docs.unity3d.com/Documentation/Manual/CommandLineArguments.html) to configure certain aspects of your game at launch. This is especially useful for dedicated server builds, where arguments let you override default network settings such as the IP address and port.
```
```suggestion
Use [command-line arguments](https://docs.unity3d.com/Documentation/Manual/CommandLineArguments.html) to configure certain aspects of your game at launch. This is especially useful for dedicated server builds, where arguments let you override default network settings such as the IP address and port.
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
```diff
@@ -0,0 +1,63 @@
+
+You can use [command line arguments](https://docs.unity3d.com/Documentation/Manual/CommandLineArguments.html) to configure certain aspects of your game at launch. This is especially useful for dedicated server builds, where arguments let you override default network settings such as the IP address and port.
+
+## Using Command Line Arguments
```
```suggestion
## Using command-line arguments
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
```diff
@@ -0,0 +1,63 @@
+
+## Using Command Line Arguments
+
+When launching a standalone build (for example, a headless dedicated server), you can supply custom arguments to modify runtime behavior.
```
```suggestion
When launching a standalone build (such as a headless dedicated server), you can supply custom arguments to modify runtime behavior.
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
```diff
@@ -0,0 +1,63 @@
+
+Reserved arguments:
+- -port
+- -ip
```
```suggestion
Reserved arguments:

- `-port`
- `-ip`
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
```diff
@@ -0,0 +1,63 @@
+
+---
+
+## Custom Arguments
```
```suggestion
## Custom arguments
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
```diff
@@ -0,0 +1,63 @@
+Use `GetArgs()` in your project code to collect and process these values.
+
+[!NOTE]
+Adding a custom command line argument requires you to explicitly retrieve and handle it in your implementation.
```
```suggestion
You can define additional custom command-line arguments and retrieve them through the `CommandLineOptions` class. Use `GetArgs()` in your project code to collect and process these values.

> [!NOTE]
> Adding a custom command-line argument requires you to explicitly retrieve and handle it in your implementation.
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
```diff
@@ -0,0 +1,63 @@
+
+---
+
+## Example: Reading Command Line Arguments
```
```suggestion
## Example

The following code shows you an example of defining and then reading a custom command-line argument.
```
Feel free to change this line if I've got the explanation wrong 😅

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
````diff
@@ -0,0 +1,63 @@
+```
+
+---
+
````
```suggestion
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
```diff
@@ -0,0 +1,63 @@
+
+---
+
+## Overriding Connection Data
```
```suggestion
## Override connection data
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
```diff
@@ -0,0 +1,63 @@
+
+## Overriding Connection Data
+
+If you want to ignore the connection **port** provided through command line arguments, you can override it by using the optional `forceOverride` parameter in:
```
```suggestion
If you want to ignore the connection port provided through command-line arguments, you can override it by using the optional `forceOverride` parameter in:
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
````diff
@@ -0,0 +1,63 @@
+UnityTransport.SetConnectionData(string ip, ushort port, string listenAddress, bool forceOverride);
+```
+
+Setting `forceOverride` to `true` ensures that the values you pass to `SetConnectionData` override any values specified via command line arguments.
````
```suggestion
Setting `forceOverride` to `true` ensures that the values you pass to `SetConnectionData` override any values specified via command-line arguments.
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/configuration.md:
```diff
@@ -2,8 +2,10 @@
+| **[Configuring connections](configure-connections.md)** | Configure connections in your project.                                                                            |
+| **[Transports](advanced-topics/transports.md)**         | Transport layers establish communication between your application and different hosts in a network.               |
+| **[Unity Relay](relay/relay.md)**                       | Use Unity Relay with Netcode for GameObjects to connect clients and hosts over the internet using a relay server. |
+| **[Command line arguments](command-line-arguments.md)** | Use command line arguments.                                                                                       |
```
```suggestion
| **[Command-line arguments](command-line-arguments.md)** | Use command-line arguments to configure certain aspects of your game at launch.                                                                              |
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/command-line-arguments.md:
````diff
@@ -0,0 +1,63 @@
+```
+
+Setting `forceOverride` to `true` ensures that the values you pass to `SetConnectionData` override any values specified via command line arguments.
+
````
```suggestion

## Additional resources

- [Command-line arguments in the Unity Manual](https://docs.unity3d.com/6000.2/Documentation/Manual/CommandLineArguments.html)
```

jabbacakes reviewed (approved):
The content is good, suggestions are mostly for consistency/formatting. Tl;dr:

- Command-line arguments needs a hyphen, so added that for all instances
- Headers should be in sentence case (capitalise only the first letter)
- Removed section breaks, we don't use those in the docs generally
- Added note formatting so that it'll render in the package docs properly

miniwolf commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -805,14 +805,60 @@ public void SetClientRelayData(string ipAddress, ushort port, byte[] allocationI
         /// <param name="listenAddress">The address the server is going to listen on.</param>
-        public void SetConnectionData(string ipv4Address, ushort port, string listenAddress = null)
+        /// <param name="forceOverrideCommandLineArgs">When true, any command line arguments will be ignored.</param>
+        public void SetConnectionData(string ipv4Address, ushort port, string listenAddress = null, bool forceOverrideCommandLineArgs = false)
```
This is a breaking change. Example: Precompiled assemblies will fail when querying the old method.

noellie-velez replied:
forceOverrideCommandLineArgs has a default value so when using SetConnectionData method as before (without the added bool) I guess it should not fail. 
That said, there is a failure in the API validation [vetting job](https://unity-ci.cds.internal.unity3d.com/job/59388928), is it related @NoelStephensUnity?
<img width="1147" height="977" alt="image" src="https://github.com/user-attachments/assets/d66cb5c5-30da-46ba-b87d-cb9f7ab7d271" />

miniwolf replied:
That is what it is complaining about. You are not allowed to change the layout of public methods. 
You can see it here in sharplab:

https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyDMABGgQMIEDeOB1RhyALAQLIAUARgPYcQECGBAXgIAzXhADOAUwCUFKjQC+86stpF0ANiKMAtrzAA7FgGVgAJ0MBzANoBdPmcvjpqythoeCByQHdSLaQA6VmkAblUlbAUgA=

The IL invocation is:
 call instance void C::M(bool) regardless of the default parameter.

miniwolf replied:
If you expand the vetting test result you also see this:

"breakingChanges": [

"MethodParameterCountChange (System.Void Unity.Netcode.Transports.UTP.UnityTransport::SetConnectionData(System.String,System.UInt16,System.String) -> System.Void Unity.Netcode.Transports.UTP.UnityTransport::SetConnectionData(System.String,System.UInt16,System.String,System.Boolean))"

]

}

noellie-velez replied:
I see, I'll fix this then. 
Thanks for the heads up and the answers!

noellie-velez committed:
Adding hyphen to command-line arguments

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Adding hyphen to command-line arguments

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Adding hyphen to command-line arguments + rewording

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Adding hyphen to command-line arguments

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Rewording

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Adding hyphen to command-line arguments + rewording

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Adding additional resources

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Adding hyphen to command-line arguments

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Adding hyphen to command-line arguments

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Styling

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Apply suggestions from docs code review

Titles in sentence case, adding a hyphen to `command-line arguments` and styling

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

noellie-velez committed:
Merge branch 'develop-2.0.0' into fix/v2.x/add-flag-dedicated-server-override

noellie-velez committed:
Avoid changing public method layout

noellie-velez committed:
Fix SetConnectionData ambiguous call

noellie-velez committed:
Init command-line options singleton when used
